snapshot script refinements and examples (#1659)
maticnetwork/matic-docs: docs/operate/snapshot-instructions-heimdall-bor.md

## Heimdall/Bor Snapshots

To begin, ensure that your node environment meets the **prerequisites** outlined [here](https://wiki.polygon.technology/docs/operate/full-node-binaries/). Before starting any services, execute the shell script provided below. This script will download and extract the snapshot data, which allows for faster bootstrapping. In our example, we will be using an Ubuntu Linux m5d.4xlarge machine with an 8TB block device attached.
To transfer the correct chaindata to your disk, follow these steps:

- When prompted, specify the network (mainnet or mumbai) and the client type (heimdall or bor).

> - The script will automatically handle the download and extraction phases, optimizing disk space by deleting already extracted files.

Reviewer: ```suggestion
> The script will automatically handle the download and extraction phases, optimizing disk space by deleting already extracted files.
```

## Heimdall/Bor Snapshots

First, set up your node with the **prerequisites** from our [PoSV1 node setup guide](https://wiki.polygon.technology/docs/operate/full-node-binaries/). Before you start any services, run the shell script below to download and extract your snapshot data for faster bootstrapping. In our example, we're using an Ubuntu Linux m5d.4xlarge machine with an 8TB Block device attached. 
User simply needs specify the proper network (mainnet/mumbai) and client type (heimdall/bor) and all the correct chaindata will start transferring to your disk. Note, this script combines download/extract phases so downloaded files that have already been extracted are deleted to minimize disk space requirements. We recommend using a Screen session to prevent any accidental interruption of chaindata download/extract script:

```
#!/bin/bash

> # ask user for network and client type

Reviewer: ```suggestion
# function to validate network input
function validate_network() {
  if [[ "$1" != "mainnet" && "$1" != "mumbai" ]]; then
    echo "Invalid network input. Please enter 'mainnet' or 'mumbai'."
    exit 1
  fi
}

# function to validate client input
function validate_client() {
  if [[ "$1" != "heimdall" && "$1" != "bor" ]]; then
    echo "Invalid client input. Please enter 'heimdall' or 'bor'."
    exit 1
  fi
}

# ask user for network and client type
```

Reviewer: if expecting a specific input for network and client, would it be worthwhile making sure the inputs are correct?

> read -p "PoSV1 Network (mainnet/mumbai): " network_input

Reviewer: ```suggestion
read -p "PoSV1 Network (mainnet/mumbai): " network_input
# validate network input
validate_network "$network_input"
```

> read -p "Client Type (heimdall/bor): " client_input

Reviewer: ```suggestion
read -p "Client Type (heimdall/bor): " client_input
# validate client input
validate_client "$client_input"
```

aria2c -x6 -s6 -i $client-$network-incremental-compiled-files.txt

# helper method to extract all files and delete already-extracted download data to minimize disk use
function extract_files() {
    compiled_files=$1
    while read -r line; do
        if [[ "$line" == checksum* ]]; then
            continue
        fi
        filename=`echo $line | awk -F/ '{print $NF}'`
        if echo "$filename" | grep -q "bulk"; then
            pv $filename | tar -I zstd -xf - -C . && rm $filename
        else
            pv $filename | tar -I zstd -xf - -C . --strip-components=3 && rm $filename
        fi
    done < $compiled_files
}

# execute final data extraction step
extract_files $client-$network-incremental-compiled-files.txt
```

**Note:** If experiencing intermittent aria2c download errors, try reducing concurrency as exampled here:
```
aria2c -c -m 0 -x6 -s6 -i heimdall-$network-incremental-compiled-files.txt --max-concurrent-downloads=1
```

After extraction, be sure to point your heimdall/bor datadir config to the path of your extracted data. 
This ensures the systemd services can properly register the snapshot data on client start. 

> Symlinks can also be used if needing to keep default client config settings.

Reviewer: ```suggestion
If you wish to preserve the default client configuration settings, you can use symbolic links (symlinks).
```

**For example:** Say you've mounted your block device at ~/snapshots and downloaded/extracted chaindata to

> 'heimdall_extract' and 'bor_extract' respectively. Using the below sample commands you can properly register

Reviewer: ```suggestion
for Heimdall into the directory `heimdall_extract`, and for Bor into the directory `bor_extract`. To ensure proper registration
```
